### PR TITLE
add TRIM support to virtual filesystem, and implementation on linux

### PR DESCRIPTION
**Pull request by jkub**
Add TRIM method to file system, which enables excising a range of the DB file. The OS can drop pages from the page-cache, and the file-system is free to deallocate the range (sparse file support). Reads to the range will succeed but will return undefined data (though most platforms will return 0s).

This is implemented and tested on linux. Windows/ntfs also has support for trim but I have not implemented it, and so tests do not exercise it.

The feature is enabled via a dbconfig.options.trim_free_blocks, which defaults to false.

### CODE REVIEW

File: src/storage/single_file_block_manager.cpp [hunk 1]
```diff
@@ -547,6 +551,28 @@ void SingleFileBlockManager::WriteHeader(DatabaseHeader header) {
 	active_header = 1 - active_header;
 	//! Ensure the header write ends up on disk
 	handle->Sync();
+	// Release the free blocks to the filesystem.
+	TrimFreeBlocks();
+	newly_freed_list.clear();
```
Comment (carlopi) [inline src/storage/single_file_block_manager.cpp:556]:
```suggestion
```
This is done already as part of TrimFreeBlocks.
Potentially to consider an D_ASSERT that `newly_freed_list is actually empty.

File: src/storage/single_file_block_manager.cpp [hunk 2]
```diff
@@ -432,9 +436,8 @@ void SingleFileBlockManager::Truncate() {
 		return;
 	}
 	// truncate the file
-	for (idx_t i = 0; i < blocks_to_truncate; i++) {
-		free_list.erase(max_block + i);
-	}
+	free_list.erase(free_list.lower_bound(max_block), free_list.end());
+	newly_freed_list.erase(newly_freed_list.lower_bound(max_block), newly_freed_list.end());
```
Comment (carlopi) [inline src/storage/single_file_block_manager.cpp:440]: I was not obvious to me looking here that both free_list and newly_freed_list are guaranteed to be sorted (given they are backed by a set<>).

Unsure if that should be made more visible.

File: test/helpers/test_helpers.cpp [hunk 3]
```diff
@@ -146,6 +146,12 @@ unique_ptr<DBConfig> GetTestConfig() {
 	result->options.checkpoint_wal_size = 0;
 #else
 	result->options.checkpoint_on_shutdown = false;
+#endif
+#if defined(__linux__)
+	// Currently this is only supported on linux.
+	// Because we test this mode on linux, and the no-trim mode on other platforms,
+	// we get coverage with the option both on and off.
+	result->options.trim_free_blocks = true;
```
Comment (carlopi) [inline test/helpers/test_helpers.cpp:154]: I am not sure about this part. I think it's great to test this thoroughly, but it looks this also means we will stop testing duckdb on Linux with its default setting, and I can see potentially some problems being visible only with one of the two modes.

Unsure what's the way forward.
Maybe leaving default to false but having some unittester options to turn this on based on environment variable?

And also, I would consider this to not be __linux__ only, and when this is turned on all filesystems handle this properly (via Trim being a no-op).

**Review by carlopi — COMMENTED** [review]
I have one substantial comment on the testing side of this, that I think might require @Mytherin to take a look and make a call on how to move forward, and then took a look also at the rest and left two minor clarifications.

**Comment by Mytherin** [comment]
Thanks for the PR! Cool functionality. I was not aware this was even possible. I'm very curious how well this works on various file systems/operating systems, but we can experiment with that separately. Could you have a look at the failing CI?

File: src/storage/single_file_block_manager.cpp (hunk 2 above)
Comment (jkub) [reply src/storage/single_file_block_manager.cpp:440]: It's part of the spec that std::set is sorted. Are you suggesting that we rename these members, to make it clearer that they're sorted?

File: test/helpers/test_helpers.cpp (hunk 3 above)
Comment (jkub) [reply test/helpers/test_helpers.cpp:154]: Thank you. I wasn't sure what to do here and was hoping for feedback.

Comment (jkub) [reply test/helpers/test_helpers.cpp:154]: I can make that change. Although DUCKDB_ALTERNATIVE_VERIFY also tests a slightly different checkpoint/shutdown behavior, which could conceivably interact?

Comment (Mytherin) [reply test/helpers/test_helpers.cpp:154]: Yes that is true, maybe putting it behind `DESTROY_UNPINNED_BLOCKS` would be better? Or `RUN_SLOW_VERIFIERS`?

Comment (jkub) [reply test/helpers/test_helpers.cpp]: Updated to use DUCKDB_RUN_SLOW_VERIFIERS

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**